lib/src/sdk/sdk_package_config.dart

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:yaml/yaml.dart';
+
+import '../utils.dart' show ExpectField, ExpectEntries;
+
+/// The top level structure of an `sdk_packages.yaml` file.
+///
+/// See https://github.com/dart-lang/pub/issues/3980 for discussion of the
+/// feature and format.
+///
+/// Version 1 of the format is as follows:
+///
+/// ```yaml
+/// # Required, the version of the format used in this file. Not all versions
+/// # will be supported forever, but some number of previous ones will be.
+/// version: 1
+///
+/// # The SDK this configuration file is targetting. Used for validation, to
+/// # ensure we are parsing a file intended for the SDK we are configuring.
+/// sdk: dart
+///
+/// # A list of package descriptors, for each package vendored by this SDK.
+/// packages:
+/// - name: my_sdk_package
+///   # A path relative to the root of the installed SDK, in URL form (with `/`
+///   # path separators).
+///   path: path/to/my_sdk_package
+/// ```
+class SdkPackageConfig {
```

Comment:
Hmm, do we really need this file at all? If we did dart-sdk/pkg/macros/ and such, those contain a pubspec.yaml, and this file is sort of just a directory listing. Maybe pub could just search dart-sdk/pkg/ and discover if it contains the named packages it's listing for? Then we don't need a whole new file format with a version integer and all of that complexity,

Reply:
@jonasfj had some reasons that he wanted another layer of indirection...

Reply:
The file gives us future possibilities like vendoring multiple versions of a package as well.

Reply:
Technically, I guess we don't need this file. We could do what we do flutter now.
Namely, we could say that everything in `$DART_ROOT/pkg/` is an SDK-package.

On the other hand, I think we should want a `$FLUTTER_ROOT/sdk_packages.yaml` file that explicitly configures what SDK-packages flutter has. That way we might stop accidentally exposing non-public APIs from Flutter (example `flutter_tools` is an SDK-package that shouldn't exist).

So given that we'd like to have `$FLUTTER_ROOT/sdk_packages.yaml`, it might be nice to be consistent and have a `$DART_ROOT/sdk_packages.yaml` as well.

I don't think we should block on this either way. I don't care too much either way.
I'd suggest we ship this.

Reply:
+1 I personally like the file as it gives a lot of future flexibility, and avoids hard coding any knowledge about the Flutter/Dart SDKs layouts into pub (other than the location of this one file).